src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java

```diff
@@ -60,217 +62,244 @@
+  public static void flushBuildsWithNoBCList() {
 
-            synchronized (buildsWithNoBCList) {
-                if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
-                    buildsWithNoBCList.clear();
-                }
+    ConcurrentHashMap<String, Build> clone = null;
+    synchronized (buildsWithNoBCList) {
+      clone = new ConcurrentHashMap<String, Build>(buildsWithNoBCList);
+    }
+    boolean anyRemoveFailures = false;
+    for (Build build : clone.values()) {
+      WorkflowJob job = getJobFromBuild(build);
+      if (job != null) {
+        try {
+          logger.info("triggering job run for previously skipped build " + build.getMetadata().getName());
+          triggerJob(job, build);
+        } catch (IOException e) {
+          logger.log(Level.WARNING, "flushBuildsWithNoBCList", e);
+        }
+        try {
+          synchronized (buildsWithNoBCList) {
+            removeBuildFromNoBCList(build);
+          }
+        } catch (Throwable t) {
+          // TODO
+          // concurrent mod exceptions are not suppose to occur
+          // with concurrent hash set; this try/catch with log
+          // and the anyRemoveFailures post processing is a bit
+          // of safety paranoia until this proves to be true
+          // over extended usage ... probably can remove at some
+          // point
+          anyRemoveFailures = true;
+          logger.log(Level.WARNING, "flushBuildsWithNoBCList", t);
+        }
+      }
 
-            }
+      synchronized (buildsWithNoBCList) {
+        if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
+          buildsWithNoBCList.clear();
         }
+
+      }
     }
+  }
+
+  static void removeBuildsForDeletedBCFromNoBCList(BuildConfig bc) {
+      String deletedBCName = bc.getMetadata().getName();
+      String deletedBCNamespace = bc.getMetadata().getNamespace();
+      ConcurrentHashMap<String, Build> clone = null;
+      synchronized (buildsWithNoBCList) {
+        clone = new ConcurrentHashMap<String, Build>(buildsWithNoBCList);
+      }
+      HashSet<String> buildsToRemove = new HashSet<String>();
+      for (Build build : clone.values()) {
+        String buildConfigName = build.getStatus().getConfig().getName();
+        String buildConfigNamespace = build.getStatus().getConfig().getNamespace();
+        if (deletedBCName.trim().equals(buildConfigName.trim()) &&
+          deletedBCNamespace.trim().equals(buildConfigNamespace.trim())) {
+            buildsToRemove.add(buildConfigNamespace + buildConfigName);
+        }
+      }
+
+      synchronized (buildsWithNoBCList) {
+          Iterator<String> iter = buildsToRemove.iterator();
+          while (iter.hasNext()) {
+              String key = iter.next();
+              logger.info("removing build " + key + " from builds with no build config list since build config " + deletedBCNamespace+ ":" + deletedBCName + " has been deleted");
+              buildsToRemove.remove(iter.next());
```

Comment:
What does this function even do? It initializes a set of builds to remove, which a string convention that doesn't seem amenable to anything (namespace + name are concatenated with no spearator). We then call `iter.next()` twice in the same loop without calling `iter.hasNext()` beforehand. We then do nothing with that set.

Reply:
I found the link where it says sequential events can be assumed:  https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L133-L136

So my edge case should not happen.  The build will get flushed by the add bc event before any delete bc event comes.

Reply:
method removed @adambkaplan